### PR TITLE
Introduce baseline types required to implement validator set updates

### PR DESCRIPTION
**Pull request by sug0**
Implements the second item of issue #242

### CODE REVIEW

**Comment by sug0** [comment]
I haven't added unit tests to these yet. I might leave that to the next PR (PrepareProposal), since I'll write the validation code then. idk, what do you think?

**Comment by james-chf** [comment]
There is one more `Cargo.lock` -  `make -C wasm_for_tests/wasm_source check` - not currently compiling

File: shared/src/proto/types.rs [hunk 1]
```diff
@@ -51,54 +52,47 @@ pub struct SignedTxData {
...
+/// to sign data in a [`Signed`] wrapper.
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+pub enum SerializeWithBorsh {}
+
+/// A generic signed data wrapper for serialize-able types.
+///
+/// The default serialization method is [`BorshSerialize`].
 #[derive(
-    Clone, Debug, BorshSerialize, BorshDeserialize, Serialize, Deserialize,
+    Eq, Clone, Debug, BorshSerialize, BorshDeserialize, Serialize, Deserialize,
```
Comment (james-chf) [inline shared/src/proto/types.rs:64]: I'm guessing `Eq` should only be derived for the case where `T: PartialEq`? It might need to be manually implemented in that case

Comment (james-chf) [reply shared/src/proto/types.rs:64]: *`T: PartialEq + Eq`

Comment (sug0) [reply shared/src/proto/types.rs:64]: `Eq` doesn't add any methods, it just means we a have a full equivalence relation.
https://doc.rust-lang.org/stable/std/cmp/trait.Eq.html

Comment (sug0) [reply shared/src/proto/types.rs:64]: if you have `PartialEq` implemented, `Eq` can be derived

Comment (james-chf) [reply shared/src/proto/types.rs:64]: I mean we can (or should?) only derive `Eq` in the case where `T: Eq`. As you can sign `T`s which don't implement `PartialEq`/`Eq`. I'm surprised the compiler lets you annotate `Eq` on `Signed<T, S>` without `T` also having the `Eq` trait bound though so I could be misunderstanding something

Comment (james-chf) [reply shared/src/proto/types.rs:64]: something like

Comment (sug0) [reply shared/src/proto/types.rs:64]: the compiler already does that automatically. it'll only implement `Eq` if the type params are also `Eq`

Comment (sug0) [reply shared/src/proto/types.rs:64]: actually not true, it'll implement it if the type params are `PartialEq`. ehhhh
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=095498c880c5653848415b9228642fb7

alright cool I'll manually implement `Eq`

File: shared/src/types/vote_extensions/validator_set_update.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+use crate::types::key::common::{self, Signature};
+use crate::types::key::{SigScheme, VerifySigError};
+
+// the namespace strings plugged into validator set hashes
+const BRIDGE_CONTRACT_NAMESPACE: &str = "bridge";
+const GOVERNANCE_CONTRACT_NAMESPACE: &str = "governance";
+
+/// Wrapper type for [`ethereum::Address`]
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct EthAddr(pub ethereum::Address);
```
Comment (batconjurer) [inline shared/src/types/vote_extensions/validator_set_update.rs:25]: We already have the type in `src/types/ethereum_events` except that we call the `to_fixed_bytes` method on the `ethereum::Address` type to get the raw bytes.

File: shared/src/types/vote_extensions/validator_set_update.rs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+            ),
+            epoch_to_token(ext.epoch),
+        ])
+    }
+
+    /// Sign this [`Vext`] with an Ethereum key.
+    ///
+    /// For more information, check the Ethereum bridge smart contract code:
+    ///   - <https://github.com/anoma/ethereum-bridge/blob/main/contracts/contract/Bridge.sol#L186>
+    pub fn new_abi_encoded(keypair: &common::SecretKey, ext: Vext) -> Self {
```
Comment (batconjurer) [inline shared/src/types/vote_extensions/validator_set_update.rs:127]: Can't we just implement a `new` method in the `types/proto.rs` defns for `Signed<T,S>`? And add an ABI marker struct there. That seemed like the whole point of that change and we get type API consistency.

File: shared/src/types/vote_extensions/validator_set_update.rs [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        let to_sign = Self::serialize_vext(&ext);
+        let sig = common::SigScheme::sign(keypair, &to_sign);
+        Self::new_from(ext, sig)
+    }
+
+    /// Verify the signature of a [`Vext`], signed by some
+    /// Ethereum key.
+    pub fn verify_abi_encoded(
```
Comment (batconjurer) [inline shared/src/types/vote_extensions/validator_set_update.rs:135]: Similary, this should just be a `verify` method in `types/proto.rs`. In fact, it might be drying to change `Signed<T,S>` so that `S` implements a trait that gives access to a serialization function. Then all of this logic only needs to be written once and we only need to implement this trait on marker structs. So it is a much more extensible design.

File: shared/src/types/vote_extensions/validator_set_update.rs [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    /// TODO: the validator's address is temporarily being included
+    /// until we're able to map a Tendermint address to a validator
+    /// address (see https://github.com/anoma/namada/issues/200)
+    pub validator_addr: Address,
+    /// The new [`Epoch`].
+    ///
+    /// Since this is a monotonically growing sequence number,
+    /// it is signed together with the rest of the data to
+    /// prevent replay attacks on validator set updates.
+    pub epoch: Epoch,
```
Comment (batconjurer) [inline shared/src/types/vote_extensions/validator_set_update.rs:166]: It also allows us to look up with valset is needed to verify signatures of user txs coming to the eth side of the bridge.

File: shared/src/types/vote_extensions/validator_set_update.rs [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+/// Provides a mapping between [`EthAddr`] and [`VotingPower`] instances.
+pub type VotingPowersMap = HashMap<EthAddr, VotingPower>;
+
+/// This trait contains additional methods for a [`HashMap`], related
+/// with validator set update vote extensions logic.
+pub trait VotingPowersMapExt {
+    /// Returns the keccak hash of this [`VotingPowersMap`]
+    /// to be signed by an Ethereum validator key.
+    fn get_bridge_hash(&self, epoch: Epoch) -> [u8; 32];
```
Comment (batconjurer) [inline shared/src/types/vote_extensions/validator_set_update.rs:187]: A super nit, but I like using the `Hash` type instead unless there is a really good reason not to. I think it just make intentions clearer in the code.

**Review by batconjurer — CHANGES_REQUESTED** [review]

File: shared/src/proto/types.rs (hunk 1 above)
Comment (james-chf) [reply shared/src/proto/types.rs]: > the compiler already does that automatically. it'll only implement `Eq` if the type params are also `Eq`

Yeah I thought this is how it should work as well and still don't fully understand this. Like for `PartialEq` the compiler could only derive it if `T: PartialEq` it seems. Maybe will revisit this.

File: shared/src/proto/types.rs [hunk 7]
```diff
@@ -51,54 +52,49 @@ pub struct SignedTxData {
     pub sig: common::Signature,
 }
 
-/// A generic signed data wrapper for Borsh encode-able data.
+/// Tag type that indicates we should use [`BorshSerialize`]
+/// to sign data in a [`Signed`] wrapper.
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+pub enum SerializeWithBorsh {}
```
[the remark this answers is not in the document]
Comment (sug0) [reply shared/src/proto/types.rs:58]: well it's a tag, we don't need to construct a value from it

**Review by batconjurer — APPROVED** [review]
LGTM!

File: shared/src/proto/types.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (batconjurer) [reply shared/src/proto/types.rs]: Good point.

File: shared/src/proto/types.rs [hunk 8]
```diff
@@ -51,54 +52,69 @@ pub struct SignedTxData {
...
+}
+
+/// A generic signed data wrapper for serialize-able types.
+///
+/// The default serialization method is [`BorshSerialize`].
 #[derive(
     Clone, Debug, BorshSerialize, BorshDeserialize, Serialize, Deserialize,
 )]
-pub struct Signed<T: BorshSerialize + BorshDeserialize> {
+pub struct Signed<T, S = SerializeWithBorsh> {
```
Comment (james-chf) [inline shared/src/proto/types.rs:86]: It'd be good if we could bound `S` here (if it's possible?) so that it's impossible to construct a `Signed` with an invalid tag, something like

```rust
pub struct Signed<T, S: SignedSerialize<T> = SerializeWithBorsh>
```

**Review by james-chf — APPROVED** [review]

File: shared/src/proto/types.rs (hunk 8 above)
Comment (sug0) [reply shared/src/proto/types.rs:86]: we already tighten `Signed` at the impl block starting on line 148. you can only verify signatures of `Signed` instances with a tag which implements `SignedSerialize`. you can't use the struct itself to instantiate new values, since it has a private field, but you can use `Signed::new_from` instead. regardless, we hit the same wall, this instance can't do jack if `S` does not implement `SignedSerialize`.

tl;dr I don't think this is an issue
